### PR TITLE
gap finder function uses gcd instead of min, adding OK and ERR

### PR DESCRIPTION
**Pull request by eftekhari-mhs**

### CODE REVIEW

File: imputation/ts_impute.go [hunk 1]
```diff
@@ -47,61 +57,130 @@ func dateGapFinder(keys []string) (string, int, int, int) {
 	case 10:
 		parseFormat = dayfmt
 	}
-
 	year, month, day := 0, 0, 0
 
-	duration := 1<<31 - 1 // A large number.
+	if parseFormat == "" {
+		return parseFormat, year, month, day, false, errors.New("The date format is not ISO 8601.")
+	}
```
Comment (jeffreyoldham) [inline imputation/ts_impute.go:64]: Please make this the default clause:

```suggestion
default:
  return parseFormat, year, month, day, false, errors.New("The date format is not ISO 8601.")
```

File: imputation/ts_impute.go [hunk 2]
```diff
@@ -47,61 +57,130 @@ func dateGapFinder(keys []string) (string, int, int, int) {
...
+				return parseFormat, year, month,
+					day, false, errors.New("The dates have multiple formats.")
+			}
 
-			year = int(y2 - y1)
-			month = int(M2 - M1)
-			day = int(d2 - d1)
+			md1 := strings.Split(keys[i], "-")
+			md2 := strings.Split(keys[i+1], "-")
+			// Check for unequal months.
```
Comment (jeffreyoldham) [inline imputation/ts_impute.go:79]: Here we should check that `len(md1) == len(md2)`.

File: imputation/ts_impute.go [hunk 3]
```diff
@@ -27,17 +28,26 @@ const (
...
 //   keys: A sorted list of times all with the same format, one of those listed above.
 // Returns:
 //   1) The time format of the keys.
-//   2) The minimum gap between the keys in terms of years, months, and days.
-// TODO(eftekhari-mbs): Modify the algorithm to compute a greatest common divisor-like algorithm, not
-// just the minimum gap.
-// TODO(eftekhari-mhs): add OK, ERR return values and handle errors.
-func dateGapFinder(keys []string) (string, int, int, int) {
+//   2) The GCD (gap between the keys) in terms of years, months, and days.
+//   3) OK if the function compute the frequency of data collection.
```
Comment (jeffreyoldham) [inline imputation/ts_impute.go:48]:
```suggestion
//   3) OK is True iff the function successfully computed the frequency of data collection.
```

File: imputation/ts_impute.go [hunk 4]
```diff
@@ -161,14 +242,24 @@ func Interpolate(ts TimeSeries, degree int) (TimeSeries, error) {
...
+	if !ok {
+		return ts, errors.New("Dates are not annual, monthly, or daily.")
+	}
 	log.Printf("Step is equal to : %v years, %v months, %v days \n", yStep, mStep, dStep)
 
-	//TODO(eftekhari-mhs): Handle errors.
-	startDate, _ := time.Parse(parseFormat, keys[0])
-	endDate, _ := time.Parse(parseFormat, keys[len(keys)-1])
+	startDate, er1 := time.Parse(parseFormat, keys[0])
+	endDate, er2 := time.Parse(parseFormat, keys[len(keys)-1])
```
Comment (jeffreyoldham) [inline imputation/ts_impute.go:258]:
```suggestion
	startDate, err1 := time.Parse(parseFormat, keys[0])
	endDate, err2 := time.Parse(parseFormat, keys[len(keys)-1])
```

File: imputation/ts_impute.go [hunk 5]
```diff
@@ -47,61 +57,130 @@ func dateGapFinder(keys []string) (string, int, int, int) {
...
-	return parseFormat, year, month, day
-}
+	if parseFormat == yearfmt || (equalMonths && equalDays) {
+		for i := 0; i < len(keys)-1; i++ {
+			if year == 1 {
+				// One year is the minimum possible gap in this case.
+				break
+			}
+			start, _ := time.Parse(parseFormat, keys[i])
+			end, _ := time.Parse(parseFormat, keys[i+1])
```
Comment (jeffreyoldham) [inline imputation/ts_impute.go:99]: We need to check for errors here and in similar code.

File: imputation/ts_impute_test.go [hunk 6]
```diff
@@ -152,6 +153,49 @@ func TestFillNA(t *testing.T) {
...
+				series: TimeSeries{
+					"2020-03-05": 42.5,
+					"2016-03-05": 18.9,
+					"2017-03-05": 28,
+				},
+				additionalKeys: []string{
+					"2019-03-05",
+					"2018-03-05",
+				},
+			},
```
Comment (jeffreyoldham) [inline imputation/ts_impute_test.go:198]: TODO(???): Add tests of parsing failures.

**Review by jeffreyoldham — COMMENTED** [review]
Thanks.

**Review by jeffreyoldham — APPROVED** [review]
Thanks.  Assuming the tests pass, let's get this submitted.  Thanks.

**Review by tjann — APPROVED** [review]

**Pull request merged**